datashader/core.py

```diff
@@ -368,10 +368,25 @@ def bypixel(source, canvas, glyph, agg):
     glyph : Glyph
     agg : Reduction
     """
-    if isinstance(source, pd.DataFrame):
-        dshape = dshape_from_pandas(source)
-    elif isinstance(source, dd.DataFrame):
-        dshape = dshape_from_dask(source)
+    # Avoid datashape.Categorical instantiation bottleneck
+    # by only retaining the necessary columns:
+    # https://github.com/bokeh/datashader/issues/396
+    if 'category' in str(source.dtypes.values):
```

Comment:
I speculate that this is the source of the slowdown, but I haven't confirmed yet. Before committing I tested a few different approaches for identifying whether a dataframe contains a categorical dtype, and this was the fastest method of the ones I tried. I may be able to get a speedup with a vectorized ufunc or a numba function; I'll try those shortly and push up the one that ends up being faster.